### PR TITLE
KAFKA-4794: Add access to OffsetStorageReader from SourceConnector

### PR DESCRIPTION
**Pull request by fhussonnois**
This a first attempt to implement Add access to OffsetStorageReader from Source Connector.
I am not sure if I did it right so I prefer to ask you your feedback. I still need to write some tests.

### CODE REVIEW

**Comment by hleb-albau** [comment]
Will this path be merged?

**Comment by fhussonnois** [comment]
Hi @rhauch, this PR has been updated regarding the [KIP-131](https://cwiki.apache.org/confluence/display/KAFKA/KIP-131+-+Add+access+to+OffsetStorageReader+from+SourceConnector)

File: connect/api/src/main/java/org/apache/kafka/connect/connector/Connector.java
Comment (rhauch) [inline connect/api/src/main/java/org/apache/kafka/connect/connector/Connector.java]: It'd be great to have JavaDoc for all API methods.

File: connect/api/src/main/java/org/apache/kafka/connect/sink/SinkConnector.java
Comment (rhauch) [inline connect/api/src/main/java/org/apache/kafka/connect/sink/SinkConnector.java]: Doesn't this need an `@Override`? Would be great to have JavaDoc here as well, since it is changing the return type defined by the inherited method.

File: connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java
Comment (rhauch) [inline connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java]: Would be great to have JavaDoc here as well, since it is changing the return type defined by the inherited method.

File: connect/api/src/main/java/org/apache/kafka/connect/sink/SinkConnectorContext.java
Comment (rhauch) [inline connect/api/src/main/java/org/apache/kafka/connect/sink/SinkConnectorContext.java]: I'd suggest a description that's a bit more straightforward and more closely mirrors the JavaDoc in the `ConnectorContext`:

> A context to allow a {@link SinkConnector} to interact with the Kafka Connect runtime.

File: connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnectorContext.java
Comment (rhauch) [inline connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnectorContext.java]: I'd suggest a description that's a bit more straightforward and more closely mirrors the JavaDoc in the `ConnectorContext`:

> A context to allow a {@link SourceConnector} to interact with the Kafka Connect runtime.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: How about `final` for the new field, and for any other fields that are set in the constructor?

**Review by rhauch — DISMISSED** [review]
Looks great, @fhussonnois! I do have a few specific requests, but all are pretty minor.

File: connect/api/src/main/java/org/apache/kafka/connect/connector/Connector.java
Comment (rhauch) [reply connect/api/src/main/java/org/apache/kafka/connect/connector/Connector.java]: Also, I wonder if it'd be better to have this be `protected` so that the connector subclass can access the context as now currently allowed by the protected field, but would prevent other objects from accessing it. We can always make it public in the future if we find we need that change, but once it's public we cannot retract it. WDYT?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (tedyu) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: Can you enclose the body of if block with curly braces ?
Same with else block.

It is easier to read.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (tedyu) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: Is it possible to trigger infinite loop:

raiseError -> reconfig -> raiseError -> reconfig ...

**Comment by fhussonnois** [comment]
Thank you @rhauch for your comments. I've updated the PR.

File: connect/api/src/main/java/org/apache/kafka/connect/connector/Connector.java
Comment (rhauch) [inline connect/api/src/main/java/org/apache/kafka/connect/connector/Connector.java:86]: Sorry to be picky, but we need an `@return` with a description here. If there were parameters, we'd also need `@param`s.

File: connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnectorContext.java
Comment (rhauch) [inline connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnectorContext.java:30]: Need a `@return` here.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: Why use the delegate? Why not just call the methods on the `WorkerConnector`'s fields from within the `SourceConnectorContext` and `SinkConnectorContext` methods? E.g.,

      @Override
      public void requestTaskReconfiguration() {
           ctx.requestTaskReconfiguration();
      }

Comment (fhussonnois) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: It is to keep the same logic as the raiseError() method. The delegate implements the ConnectorContext.

Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: ok

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: I know this line hasn't been changed, but the `this` variable refers to the `ConnectorContext` that does not have a `toString()` method. Might be better to refer to `WorkerConnector.this` instead, so that it behaves like old line 87.

**Comment by fhussonnois** [comment]
@rhauch  is there anything I should add or modify for the PR ?

**Comment by rhauch** [comment]
@fhussonnois, I just [submitted the KIP for a vote](https://www.mail-archive.com/dev@kafka.apache.org/msg79679.html), which will be open for a few days. This can be merged once that is approved.

**Comment by frankcoutinho** [comment]
I also needed this feature. #feelsbad

**Comment by rhauch** [comment]
@fhussonnois, would you mind updating this PR to eliminate the conflicts?

Also, while I hope to get this into AK 2.6, the KIP hasn't yet passed. I did make a suggestion on the vote thread, though, to consider making it much easier for connectors that want to also run in older versions of Connect to easily get the OffsetStorageReader when Connect supports it without jumping through a lot of ugly hoops.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: How about we consider using three nested classes to make this logic a bit easier to read?

**Comment by rhauch** [comment]
ok to test

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: At one point these were anonymous classes, and the delegation perhaps made a bit more sense. Now that they are inner classes, it seems like it would be a lot less confusing and simpler if the `DelegateToWorkerConnectorContext` class were extended by the `DelegateSinkConnectorContext` and `DelegateSourceConnectorContext` classes. Doing this has several advantages:
1. removes nesting/delegation and eliminates the chance of getting into an infinite loop
2. eliminates duplicate code in these classes
3. simplifies the context classes quite a bit
4. makes it more obvious that the sink connector context has nothing extra over the base
5. makes it more obvious that the source connector context only adds the offset storage reader
6. simplifies this code a bit (see below)

By keeping `DelegateToWorkerConnectorContext` class non-abstract, we're actually able to maintain backward compatibility by calling initialize when the connector class is neither a source or sink connector:

This code becomes simpler, too:
```suggestion
            if (isSinkConnector()) {
                SinkConnectorConfig.validate(config);
                connector.initialize(new DelegateSinkConnectorContext());
            } else if (isSourceConnector()) {
                connector.initialize(new DelegateSourceConnectorContext(offsetStorageReader));
            } else {
                connector.initialize(new DelegateToWorkerConnectorContext());
            }
```
This seems a little strange, but we're actually not checking whether the `Connector` instance passed into this `WorkerConnector` is only an instance of `SourceConnector` or `SinkConnector`. If it is neither, prior to this change the framework would still initialize it, and I think we should maintain that arguably strange behavior just to maintain backward compatibility.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Nit: let's avoid unrelated line additions.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: Nit: let's avoid unrelated line additions.

File: connect/api/src/main/java/org/apache/kafka/connect/sink/SinkConnector.java
Comment (rhauch) [inline connect/api/src/main/java/org/apache/kafka/connect/sink/SinkConnector.java:38]: Right now, this method is unused, which means we have no tests for this. The existing `MockConnector` class uses the `context` field from `Connector`, and at a minimum I think it be worthwhile to consider changing that to call the `context()` method. This class is wrapped by `MockSinkConnector` and `MockSourceConnector`, and used in unit tests.

However, it's probably worth adding a few new unit test classes for `SinkConnector` and `SourceConnector` so that we can verify things like `context()` returns the same instance as supplied to `initialize(...)`, that the `context` field also has the same instance as supplied to `initialize(...)`, and that the casting in `SinkConnector.context()` and `SourceConnector.context()` is actually safe.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java
Comment (rhauch) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java:100]: So this existing test is a little strange because it does not instantiate a source or sink connector but merely relies upon the WorkerConnector for the most part has the same behavior regardless of whether the `Connector` instance is an instance of `SourceConnector` or `SinkConnector`. This PR changes this behavior, and I think we need to add a bit more testing here.

I'd suggest adding three more tests that simply verify the correct context type is passed to the `connector.initialize(...)` method, and that these tests verify the context is:
* a `ConnectorContext` instance if the connector is neither a source or a sink
* a `SinkConnectorContext` instance if the connector is a sink
* a `SourceConnectorContext` instance if the connector is a source

**Review by rhauch — COMMENTED** [review]
Thanks, @fhussonnois. I have a few more suggestions to clean up the implementation a bit more, to improve backward compatibility, and to ensure the new code is adequately tested.

**Comment by rhauch** [comment]
@fhussonnois, thanks again for doing the brunt of the work on this feature. We're getting close to AK 2.6.0 feature freeze, and it'd be great to merge this feature now that KIP-131 has been adopted. Since my previous review was some time ago, I hope you don't mind that I pushed the remaining changes I suggested in my previous review. I've also dismissed my previous "Request Changes" review.

@kkonstantine would you mind taking a look, since I'd rather have an independent review of my changes? Thanks!

**Comment by rhauch** [comment]
retest this please

**Comment by rhauch** [comment]
ok to test

**Review by rhauch — APPROVED** [review]
@fhussonnois, thanks again for doing the brunt of the work on this feature. We're getting close to AK 2.6.0 feature freeze, and it'd be great to merge this feature now that KIP-131 has been adopted. Since my previous review was some time ago, I hope you don't mind that I pushed the remaining changes I suggested in my previous review. I've also dismissed my previous "Request Changes" review.

@kkonstantine would you mind taking a look? I'd like to have an independent review of my changes. Thanks!

**Comment by fhussonnois** [comment]
@rhauch thank you very much for finalizing this PR; I apologize for not finding any free time to work on it these past few weeks.

**Comment by rhauch** [comment]
retest this please

**Comment by rhauch** [comment]
Just one failure in each of the 3 builds, and none were in Connect unit or integration tests.

File: connect/api/src/test/java/org/apache/kafka/connect/connector/ConnectorTest.java
Comment (kkonstantine) [inline connect/api/src/test/java/org/apache/kafka/connect/connector/ConnectorTest.java]: Should we remove this class? Currently they can only be 2 types of connectors, source or sink connectors. This seems to introduce a third implementation of the `abstract Connector` class, but testing seems sufficient without it.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: adding `final` in every argument here was just noise but with not much added value. Given that we don't write constructors like this in Connect I think we should revert and add the new argument without the `final` specifier.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: We don't allow a third option [elsewhere](https://github.com/apache/kafka/blob/trunk/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/distributed/DistributedHerder.java#L1288-L1292) in the code, so probably we shouldn't here either.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (kkonstantine) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]:
```suggestion
        WorkerSourceConnectorContext(OffsetStorageReader offsetStorageReader) {
```

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java
Comment (kkonstantine) [inline connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java]: I'm not convinced we should allow this

File: connect/api/src/main/java/org/apache/kafka/connect/sink/SinkConnector.java
Comment (kkonstantine) [inline connect/api/src/main/java/org/apache/kafka/connect/sink/SinkConnector.java]:
```suggestion
```
we are not adding anything to the javadoc of the super method, so we shouldn't need `@inheritDoc`

File: connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java
Comment (kkonstantine) [inline connect/api/src/main/java/org/apache/kafka/connect/source/SourceConnector.java]:
```suggestion
```
we are not adding anything to the javadoc of the super method, so we shouldn't need `@inheritDoc`

**Review by kkonstantine — COMMENTED** [review]
Thanks for the original contribution and the KIP @fhussonnois 

@rhauch thanks for completing this PR. Overall looks good. I have a few minor comments and then I think that maybe we shouldn't plan for a connector implementation that is not a sink or a source in the tests. The code would break elsewhere anyways, and this is not supported by they framework today. Wdyt?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java]: Good call. I'll update the code to throw an exception like in `Worker` when creating a source and sink task.

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java]: Changing the test to expect a failure.

**Comment by rhauch** [comment]
Rebased to correct conflicts.

**Comment by rhauch** [comment]
retest this please

**Comment by rhauch** [comment]
ok to test

**Comment by rhauch** [comment]
retest this please

File: connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java
Comment (rhauch) [reply connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerConnectorTest.java:100]: Added more tests and changed existing tests to use source or sink and not an ambiguous connector.

**Comment by rhauch** [comment]
All 3 builds had test failures on known flaky Streams integration tests, but all Connect-related tests passed.

**Review by kkonstantine — APPROVED** [review]
This looks great now. Happy to see this feature being added! 
Thanks @rhauch 
LGTM

**Pull request merged**
